doc/index.md

```diff
@@ -1,61 +1,119 @@
+-  HTTP 1 and HTTP 2 as specified by [RFC 7230](https://tools.ietf.org/html/rfc7230) and [RFC 7540](https://tools.ietf.org/html/rfc7540)
+-  WebSockets as specified by [RFC 6455](https://tools.ietf.org/html/rfc6455) and [RFC 7936](https://tools.ietf.org/html/rfc7936) including ping/pong, binary data transfer and TLS encyrption
+-  Transport Layer Security (TLS)- lua-http supports HTTPS and WSS via [luaossl](https://github.com/wahern/luaossl).  
+-  Fully asynchronous API that does not block the current thread when executing operations that typically block
+-  Easy integration into other event-loop based application models
 
+lua-http was written to fill a gap in the Lua ecosystem by providing an asynchrounous HTTP and WebSocket capabilities to Lua applications without forcing the developer to follow a specific pattern, or even know anything about the transportation of HTTP data (other than the connection addresses). The library is simple and un-obtrusive and is designed to accommodate tens of thousands of connections on commodity hardware. This allows other developers to concentrate on line-of-business features when building Internet enabled applications, instead of worrying about the details of connection management. 
 
-## Conventions
 
-Operations that may block the current coroutine take an optional timeout.
+## Common use cases
 
-HTTP 1 request and status line fields are passed around inside of [headers](#http.headers) objects under keys `":authority"`, `":method"`, `":path"`, `":scheme"` and `":status"` as defined in HTTP 2. As such, they are all kept in string form (important to remember for the `:status` field).
+The following are a few basic examples of how the lua-http library can be used. 
 
-Header fields should always be used with lower case keys.
+###Retrieving a document
 
-### Errors
+The highest level interface for clients is [*http.request*](#http.request). By constructing a [*request*](#http.request) object from a uri using [`new_from_uri`](#http.request.new_from_uri) and immediately evaluating it, you can easily fetch an HTTP resource.
 
-Invalid function parameters will throw a lua error (if validated).
+```lua
+local http_request = require "http.request"
+local headers, stream = assert(http_request.new_from_uri("http://example.com"):go())
+local body = assert(stream:get_body_as_string())
+if headers:get ":status" ~= "200" then
+	error(body)
+end
+print(body)
+```
 
-Errors are returned as `nil, error, errno` unless noted otherwise.
+###Connecting To A WebSocket Server
 
-Some HTTP 2 operations return/throw special [http 2 error objects](#http.h2_error).
+lua-http WebSocket client connections can be used within the cqueues library to send data and wait for incoming data.
 
+```lua
+local cqueues = require "cqueues"
 
-## Terminology
+local websocket = require "http.websocket"
 
-Much lua-http terminology is borrowed from HTTP 2.
+local uri = [[ws://my.websocketserver.com/path:8000]]
```

Comment:
why lua "long" style string?

The ':8000' at the end is weird, is it meant to be a port (if so it goes before the `/path`)